File: draughts/cpp/minimax.cpp
#include "minimax.h"

#include <time.h>
#include <stdlib.h>

Minimax::Minimax(int depth, bool first_player, int option /*= 1*/){
  this->depth = depth;
  this->first_player = first_player;
  this->option = option;
  srand(42);
}

void Minimax::init(Board *board){
  this->board = board;
  this->player = this->first_player;
}

void Minimax::play(Play p){
  player = !player;
}

//best move saved in board
void Minimax::search(){

  auto plays = board->getPlays(player);
  board->best_play = plays[0];

  //[TODO] test if wins next move

  int alfa = INT_MIN;
  int beta = INT_MAX;

  if(player) //if we are player1, we want to maximize the board score
  max_value(board,alfa,beta,depth);
  else //if we are not player1, we want to minimize the board score
  min_value(board,alfa,beta,depth);
}

int Minimax::max_value(Board *board, int alfa, int beta, int depth_max)
{
  if(board->gameOver(true)){
    if(board->whoWins(true)) return -1000000;
    return 0;
  }

  if (board->depth >= depth_max) return board->eval_board(option);

  int val = INT_MIN, valx;
  int podar = 0;

  auto plays = board->getPlays(true);

  std::random_shuffle(plays.begin(), plays.end());

  for(Play p : plays){ if(podar) break;

    //play
    board->play(p);

    //call min
    board->depth++;
    valx = min_value(board, alfa, beta, depth_max);
    board->depth--;

    //rmplay
    board->rmplay();

    //process valx
    if (val < valx){
      if (board->depth == 0) board->best_play = p;
      val = valx;
    }

    if (val>=beta){
      podar=1;
    }

    alfa = std::max (alfa,val);
  }

  return val;
}

int Minimax::min_value(Board *board, int alfa, int beta, int depth_max)
{
  if(board->gameOver(false)){
    if(board->whoWins(false)) return 1000000;
    return 0;
  }
  if ( board->depth >= depth_max) return board->eval_board(option);

  int val = INT_MAX, valx;
  int podar = 0;

  auto plays = board->getPlays(false);

  std::random_shuffle(plays.begin(),plays.end());

  for(Play p : plays){ if(podar) break;
    //play
    board->play(p);

    //call min
    board->depth++;
    valx = max_value(board, alfa, beta, depth_max);
    board->depth--;

    //rmplay
    board->rmplay();

    //process valx
    if (val > valx){
      if (board->depth == 0) board->best_play = p;
      val = valx;
    }

    if (val<=alfa){
      podar=1;
    }

    beta = std::min (beta,val);
  }

  return val;
}

File: draughts/python/minimax.py
from play import *
from board import *

import random

class Minimax:
    __slots__ = ['depth', 'first_player', 'board', 'player']

    def __init__(self, depth, first_player):
        self.depth = depth
        self.first_player = first_player
        random.seed(42)
        return

    def init(self, board):
        self.board = board
        self.player = self.first_player
        return

    def play(self, p):
        self.player = not self.player
        return

    def search(self):
        plays = self.board.getPlays(self.player)
        self.board.best_play = plays[0]

        alfa = -10000000
        beta =  10000000

        if self.player:
            self.max_value(alfa, beta)
        else:
            self.min_value(alfa, beta)
        return

    def max_value(self, alfa, beta):
        if self.board.gameOver(True) :
            if self.board.whoWins(True): return -10000
            return 0

        if self.board.depth >= self.depth: return self.board.eval_board()

        val = -10000000
        valx = 0
        podar = False

        plays = self.board.getPlays(True)
        random.shuffle(plays)

        for p in plays:
            if podar: break

            self.board.play(p)
            self.board.depth += 1
            valx = self.min_value(alfa, beta)
            self.board.depth -= 1
            self.board.rmplay()

            if val < valx:
                if self.board.depth == 0: self.board.best_play = p
                val = valx

            if val >= beta: podar = True
            alfa = max(alfa,val)

        return val

    def min_value(self, alfa, beta):
        if self.board.gameOver(False) :
            if self.board.whoWins(False): return 10000
            return 0

        if self.board.depth >= self.depth: return self.board.eval_board()

        val = 10000000
        valx = 0
        podar = False

        plays = self.board.getPlays(False)
        random.shuffle(plays)

        for p in plays:
            if podar: break

            self.board.play(p)
            self.board.depth += 1
            valx = self.max_value(alfa, beta)
            self.board.depth -= 1
            self.board.rmplay()

            if val > valx:
                if self.board.depth == 0: self.board.best_play = p
                val = valx

            if val <= alfa: podar = True
            beta = min(beta,val)

        return val
